### PR TITLE
Message signer implementation

### PR DESCRIPTION
**Pull request by samphillips-f3**
This provides interfaces for signing HTTP requests as per the [HTTP Message Signing RFC](https://www.ietf.org/archive/id/draft-cavage-http-signatures-12.txt).

The initial implementation provides:
* HTTP Request signing using `rsa-sha256` or `rsa-sha512`
* Digest header generation using `SHA-256` or `SHA-512`
* Signature population on either `Authorization` or `Signature` header
* Support for the `(request-target)` special case signature string header

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,52 @@
-# go-http-message-signatures
-Go implementation of the HTTP Message Signatures RFC
+# Go HTTP Message Signatures
+
+Go implementation of the [HTTP Message Signatures RFC](https://www.ietf.org/archive/id/draft-cavage-http-signatures-12.txt).
```
Comment (jlevconoks-form3) [inline README.md]: I would add the version note here, similar to the one you have in doc.go

**Review by jlevconoks-form3 — DISMISSED** [review]

**Review by hubertbarc-form3 — DISMISSED** [review]

**Review by liam-shepherd-form3 — DISMISSED** [review]

File: message_signer.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+// SignRequest method signs provided http.Request
+// signatureHeaders is a list of header names that specifies which of them (together with their values) will be signed. At least one is required.
+func (ms *MessageSigner) SignRequest(req *http.Request, signatureHeaders []string) (*http.Request, error) {
+	if len(signatureHeaders) == 0 {
+		return nil, NewValidationError("no signature headers set, at least one should be specified")
+	}
+
+	body, err := readBody(req)
+	if err == nil && len(body) != 0 {
```
Comment (vu-hoang-phan-form3) [inline message_signer.go]: I think the error is ignored here if not empty, shouldn't we return the error in case reading the body fails?

Comment (samphillips-f3) [reply message_signer.go]: good catch

File: .goreleaser.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  use: git
+  sort: asc
+
+  groups:
+    - title: Breaking Changes
+      regexp: "^.*BREAKING CHANGE[(\\w)]*:+.*$"
+      order: 0
+    - title: Features
+      regexp: "^.*feat[(\\w)]*:+.*$"
+      order: 0
```
Comment (njasm-f3) [inline .goreleaser.yaml]: Question: Shouldn't each group have it's unique order index here?

Comment (samphillips-f3) [reply .goreleaser.yaml]: yes, copy/paste error I think

File: .github/workflows/master.yaml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        with:
+          go-version-file: 'go.mod'
+      - name: Run goreleaser release
+        uses: goreleaser/goreleaser-action@b953231f81b8dfd023c58e0854a721e35037f28b # v2.9.1
+        with:
+          distribution: goreleaser
+          version: v1.7.0
+          args: release --rm-dist
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
Comment (njasm-f3) [inline .github/workflows/master.yaml]:
```suggestion
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}

```

File: .github/dependabot.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+version: 2
+updates:
+  - package-ecosystem: "gomod"
+    directory: "/"
+    schedule:
+        interval: weekly
+        time: "08:00"
+        timezone: Europe/London
+    pull-request-branch-name:
+        separator: "-"
```
Comment (njasm-f3) [inline .github/dependabot.yaml]:
```suggestion
        separator: "-"

```

File: .gitignore [hunk 6]
```diff
@@ -0,0 +1,3 @@
+.vscode/
+
+tools/
```
Comment (njasm-f3) [inline .gitignore]:
```suggestion
tools/

```

File: .golangci.yaml [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    camel_name: false
+
+issues:
+  exclude-rules:
+    - path: _test\.go
+      linters:
+        - stylecheck
+        - goerr113
+        - lll
+        - funlen
```
Comment (njasm-f3) [inline .golangci.yaml]:
```suggestion
        - funlen

```

File: .goreleaser.yaml [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+      regexp: "^.*(chore|build|style|refactor|perf|test)[(\\w)]*:+.*$"
+      order: 3
+    - title: Others
+      order: 999
+
+  filters:
+    exclude:
+      - '^docs'
+      - '^ci'
+      - typo
```
Comment (njasm-f3) [inline .goreleaser.yaml]:
```suggestion
      - typo

```

File: Makefile [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+.PHONY: tools/golangci-lint
+tools/golangci-lint:
+	@echo "==> Installing golangci-lint..."
+	@./scripts/install-golangci-lint.sh $(GOLANGCI_VERSION)
+
+.PHONY: tools/goimports
+tools/goimports:
+	@echo "==> Installing goimports..."
+	@GOBIN=$$(pwd)/tools/ go install golang.org/x/tools/cmd/goimports@$(GOIMPORTS_VERSION)
```
Comment (njasm-f3) [inline Makefile]:
```suggestion
	@GOBIN=$$(pwd)/tools/ go install golang.org/x/tools/cmd/goimports@$(GOIMPORTS_VERSION)

```

File: scripts/calculate-next-version.sh [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    echo >&2 "You can rebase your branch to fix your commits and then force push again."
+    exit 1
+  fi
+fi
+
+FULL_TAG="v${NEXT_SEMVER_TAG}"
+
+echo "Version change ${PREV_SEMVER_TAG} -> ${FULL_TAG}"
+
+export FULL_TAG
```
Comment (njasm-f3) [inline scripts/calculate-next-version.sh]:
```suggestion
export FULL_TAG

```

File: errors.go [hunk 11]
```diff
@@ -0,0 +1,117 @@
+package httpsignatures
+
+import "fmt"
+
+type HashingError struct {
```
Comment (piotrolchawa-f3) [inline errors.go:6]: Should we document the errors we expose?

File: signer.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+	case crypto.SHA256:
+		return &HashingAlgorithm{algo, sha256Name, sha256.New}, nil
+	case crypto.SHA512:
+		return &HashingAlgorithm{algo, sha512Name, sha512.New}, nil
+	default:
+		return nil, NewInitialisationError(fmt.Sprintf("crypto hash is unknown or unsupported: %s", algo), nil)
+	}
+}
+
+// Signer interface describes something that can sign content.
```
Comment (piotrolchawa-f3) [inline signer.go]:
```suggestion
// Signer is the interface that wraps the basic Sign method.

// Sign signs given content using rand as a good source of entropy for blinding the signing operation. It returns the generated signature and any error encountered that caused sign to stop early. Sign must return a non-nil error if it cannot properly generate the requested signature. Sign must not modify the slice content, even temporarily.
```

I used [io.Writer](https://pkg.go.dev/io#Writer) interface docs as a template.

**Review by piotrolchawa-f3 — CHANGES_REQUESTED** [review]

File: errors.go (hunk 11 above)
Comment (samphillips-f3) [reply errors.go:6]: done

File: README.md [hunk 13]
```diff
@@ -1,2 +1,54 @@
...
-Go implementation of the HTTP Message Signatures RFC
+# Go HTTP Message Signatures
+
+Go implementation of the [draft HTTP Message Signatures RFC](https://www.ietf.org/archive/id/draft-cavage-http-signatures-12.txt). 
+
+**Please note the version of the draft this is based on as there are varying versions which are incompatible.**
+
+## Usage
+
+`import "github.com/form3tech-oss/go-http-message-signatures"
```
Comment (jlevconoks-form3) [inline README.md:9]: nit: I think it is more common to write `go get ... ` in readme instead of `import`

**Review by piotrolchawa-f3 — DISMISSED** [review]

**Review by piotrolchawa-f3 — APPROVED** [review]

**Comment by hubertbarc-form3** [comment]
Approved.

**Pull request merged**